### ISSUE TITLE
Truncate links managing visibility of spring vectors and forces

### ISSUE THREAD

**Issue by Denz1994** [phetsims/masses-and-springs#71]
Current workable code looks like this:

```js
    Property.multilink( [ mass.springProperty, model.velocityVectorVisibilityProperty, mass.userControlledProperty ],
      function( springMassAttachedTo, visible, userControlled ) {
        if ( springMassAttachedTo !== null && visible === true && userControlled === false ) {self.velocityArrow.visible = visible;}
        else if ( springMassAttachedTo === null || visible === false || userControlled === true ) {self.velocityArrow.visible = false;}
      } );

    Property.multilink( [ mass.springProperty, model.accelerationVectorVisibilityProperty, mass.userControlledProperty ],
      function( springMassAttachedTo, visible, userControlled ) {
        if ( springMassAttachedTo !== null && visible === true && userControlled === false ) {
          self.accelerationArrow.visible = visible;
        }
        else if ( springMassAttachedTo === null || visible === false || userControlled === true ) {
          self.accelerationArrow.visible = false;
        }
      } );

    Property.multilink( [ mass.springProperty, model.gravityVectorVisibilityProperty, model.forcesModeProperty ],
      function( springMassAttachedTo, visible, forcesVisible ) {
        if ( springMassAttachedTo !== null && visible === true && forcesVisible === 'forces' ) {
          self.gravityForceArrow.visible = visible;
        }
        else if ( springMassAttachedTo === null || visible === false || forcesVisible === 'netForce' ) {
          self.gravityForceArrow.visible = false;
        }
      } );

    Property.multilink( [ mass.springProperty, model.springVectorVisibilityProperty, model.forcesModeProperty ],
      function( springMassAttachedTo, visible, forcesVisible ) {
        if ( springMassAttachedTo !== null && visible === true && forcesVisible === 'forces' ) {
          self.springForceArrow.visible = visible;
        }
        else if ( springMassAttachedTo === null || visible === false || forcesVisible === 'netForce' ) {
          self.springForceArrow.visible = false;
        }
      } );

    Property.multilink( [ mass.springProperty, model.netForceVectorVisibilityProperty, model.forcesModeProperty ],
      function( springMassAttachedTo, visible, forcesVisible ) {
        if ( springMassAttachedTo !== null && forcesVisible === 'netForce' ) {
          self.netForceArrow.visible = true;
        }
        else if ( springMassAttachedTo === null || visible === false || forcesVisible === 'forces' ) {
          self.netForceArrow.visible = false;
        }
      } );

    Property.multilink( [ mass.springProperty, model.gravityVectorVisibilityProperty, model.springVectorVisibilityProperty, model.forcesModeProperty ],
      function( springMassAttachedTo, gravityForceVisible, springForceVisible, forcesVisible ) {
        forceNullLine.visible =
          (springMassAttachedTo !== null && ( gravityForceVisible || springForceVisible || (forcesVisible === 'netForce') ) ) ?
          true : false;
      } );
```

I believe this can be refactored.

**Comment by Denz1994** [comment]
I was able to remove about 40 lines of code and it does look cleaner and concise, but can we do better?

Current code:
```js
    // Link handling the visibility of the velocity vector
    Property.multilink( [
      mass.springProperty,
      model.velocityVectorVisibilityProperty,
      model.accelerationVectorVisibilityProperty,
      mass.userControlledProperty
    ], function( massAttached, vectorVisible, accelerationVisible, userControlled ) {
      self.velocityArrow.visible = massAttached !== null && vectorVisible === true && userControlled === false;
      self.accelerationArrow.visible = massAttached !== null && accelerationVisible === true && userControlled === false;
    } );

    // Links handling the visibility of gravity vector
    Property.multilink( [ mass.springProperty,
      model.gravityVectorVisibilityProperty,
      model.springVectorVisibilityProperty,
      model.forcesModeProperty
    ], function( massAttached, gravityVisible, springForceVisible, forcesVisible ) {
      self.springForceArrow.visible = ( massAttached !== null && springForceVisible === true && forcesVisible === 'forces' );
      self.gravityForceArrow.visible = ( massAttached !== null && gravityVisible === true && forcesVisible === 'forces' );
      self.netForceArrow.visible = ( massAttached !== null && forcesVisible === 'netForce' );
    } );

    // Links handling the visibility of line at base of vectors
    mass.springProperty.link( function( massAttached ) {forceNullLine.visible = massAttached !== null} );
```

**Comment by Denz1994** [comment]
Tagging @samreid for a review perhaps on Thursday.

**Comment by samreid** [comment]
Thanks, it is looking much better. I have a few preliminary recommendations:

1. Eliminate `===true` and `!==null` and use truthiness instead
2. eliminate extra ()
3. name the callback parameters to match the properties
4. Factor out a constant for `'forces'`

I believe after this is complete, moving back to the "one arrow per multilink" will be better (making it clear what depends on what), we'll have to take a look.

**Comment by Denz1994** [comment]
I made those adjustments. Thanks for the suggestions.

**Comment by samreid** [comment]
Thanks, I'll try factoring them out momentarily.

**Comment by samreid** [comment]
Also, is `!!` necessary?  If so, why?

**Comment by samreid** [comment]
I was thinking of something like this:

```js
    // Show/hide the velocity arrow
    Property.multilink( [ mass.springProperty, model.velocityVectorVisibilityProperty, mass.userControlledProperty ],
      function( spring, velocityVectorVisibility, accelerationVectorVisibility, userControlled ) {
        self.velocityArrow.visible = spring && velocityVectorVisibility && !userControlled;
      }
    );

    // Show/hide the acceleration arrow
    Property.multilink( [ mass.springProperty, model.accelerationVectorVisibilityProperty, mass.userControlledProperty ],
      function( spring, velocityVectorVisibility, accelerationVectorVisibility, userControlled ) {
        self.accelerationArrow.visible = spring && accelerationVectorVisibility && !userControlled;
      }
    );

    // Show/hide the spring force arrow
    Property.multilink( [ mass.springProperty, model.springVectorVisibilityProperty, model.forcesModeProperty ],
      function( spring, springVectorVisibility, forcesMode ) {
        self.springForceArrow.visible = spring && springVectorVisibility && forcesMode === 'forces';
      }
    );

    // Show/hide the gravity force arrow
    Property.multilink( [ mass.springProperty, model.gravityVectorVisibilityProperty, model.forcesModeProperty ],
      function( spring, gravityVectorVisibility, forcesMode ) {
        self.gravityForceArrow.visible = spring && gravityVectorVisibility && forcesMode === 'forces';
      }
    );

    // Show/hide the net force arrow
    Property.multilink( [ mass.springProperty, model.forcesModeProperty ],
      function( spring, forcesMode ) {
        self.netForceArrow.visible = spring && forcesMode === 'netForce';
      }
    );
```

**Comment by Denz1994** [comment]
Oh, I see. I misunderstood.

I was following documentation from this link for your first point.
https://www.sitepoint.com/javascript-truthy-falsy/

**Comment by Denz1994** [comment]
If I don't add ```!!spring``` and leave it with either ```!spring``` or ```spring``` then I get an error. The value for spring is null when there isn't a spring attached and the vectors should only be visible when they are attached to a spring. Is this just a syntactical error?

Is there a shorthand for ```spring !== null```?

**Comment by samreid** [comment]
Thanks for helping me understand the problem.  It looks like Node.setVisible() requires a boolean, not just a truthy value, so we will need to use `!!spring` as you recommended.

**Comment by Denz1994** [comment]
Okay that makes sense then.

**Comment by samreid** [comment]
- [x] also check to see if the other callbacks can be factored out like that

**Comment by Denz1994** [comment]
Fixed in commit b759d77506476953ca4dbc5fd0796fbfc240630e
